### ISSUE TITLE
Silent data loss regression: Updating a subdocument to the same value 2 or more times loses update

### ISSUE THREAD

**Issue by johnpeb** [Automattic/mongoose#11395]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

Updating a nested object from null to the same value 2 or more times does not save the update

**If the current behavior is a bug, please provide the steps to reproduce.**

```
$ npx ts-node scripts/mongoose.ts
Got: undefined, expected: 123
```

Here's the script: 

```
import { Schema } from 'mongoose';
import * as mongoose from 'mongoose'

run().catch(err => console.log(err));

async function run(): Promise<void> {
  await mongoose.connect('mongodb://localhost:27017/test-mongoose', {});
  await mongoose.connection.dropDatabase();

  const Order = mongoose.model('Order', new Schema({
    co: { value: Number }
  }));

  await Order.create({});

  let doc = await Order.findOne();
  doc.co = {value: 123}
  doc.co = {value: 123} // < If this line is not present, there is no problem
  await doc.save();

  const res = await Order.findById(doc._id);
  console.log(`Got: ${res.co.value}, expected: 123`);
  process.exit()
}

```

**What is the expected behavior?**

`res.co.value` should be `123`

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

mongoose: 6.2.1
node: 16.13.1
mongo v4.2.8

**Comment by johnpeb** [comment]
This is a regression - it works in 5.13.13, doesn't work in 6.2.1

**Comment by Uzlopak** [comment]
You could/should pass an interface into model

```typescript
import { Schema } from 'mongoose';
import * as mongoose from 'mongoose'

interface IOrder {
    co: { value: number };
}

run().catch(err => console.log(err));

async function run(): Promise<void> {
    await mongoose.connect('mongodb://localhost:27017/test-mongoose', {});
    await mongoose.connection.dropDatabase();

    const Order = mongoose.model<IOrder>('Order', new Schema({
        co: { value: Number }
    }));

    await Order.create({});

    let doc = await Order.findOne();
    doc.co = { value: 123 }
    doc.co = { value: 123 } // < If this line is not present, there is no problem
    await doc.save();

    const res = await Order.findById(doc._id);
    console.log(`Got: ${res.co.value}, expected: 123`);
    process.exit()
}

```

**Comment by johnpeb** [comment]
@Uzlopak the problem is the silent data loss caused by faulty change detection logic. This is not a typescript issue. I'll update the title to be more clear.

**Comment by johnpeb** [comment]
Here's a variation that I think is the same issue:

```
import { Schema } from 'mongoose';
import * as mongoose from 'mongoose'

run().catch(err => console.log(err));

async function run(): Promise<void> {
  await mongoose.connect('mongodb://localhost:27017/test-mongoose', {});
  await mongoose.connection.dropDatabase();

  const Order = mongoose.model('Order', new Schema({
    co: { value: Number }
  }));

  await Order.create({});

  let doc = await Order.findOne();
  doc.co.value = 123
  doc.co = doc.co // < If this line is not present, there is no problem
  await doc.save();

  const res = await Order.findById(doc._id);
  console.log(`Got: ${res.co.value}, expected: 123`);
  process.exit()
}
```

**Comment by IslandRhythms** [comment]
```javascript
const mongoose = require('mongoose');
const {Schema} = mongoose;

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test-mongoose', {});
  await mongoose.connection.dropDatabase();

  const Order = mongoose.model('Order', new Schema({
    co: { value: Number }
  }));

  await Order.create({});

  let doc = await Order.findOne();
  doc.co.value = 123
  doc.co = doc.co // < If this line is not present, there is no problem
  await doc.save();

  const res = await Order.findById(doc._id);
  console.log(`Got: ${res.co.value}, expected: 123`);
  process.exit()
}
```

**Comment by vkarpov15** [comment]
It looks like this was fixed with #11428 in v6.2.4. Please upgrade to 6.2.4 for the fix.

We added a test case to make sure this doesn't happen again. Sorry for the issue.

**Comment by johnpeb** [comment]
Thanks, @vkarpov15! I can confirm it's working for us.
